rfcs/001-microsite-v2.md

```diff
@@ -0,0 +1,93 @@
+
+## JSX Compiler
+
+The core of Microsite v2.0.0 would be built around a custom JSX compiler. Conceptually, this would be similar to [Solid](https://www.solidjs.com/)'s approach with [`dom-expressions`](http://npm.im/dom-expressions)—leverage JSX/TSX as the defacto standard for authoring Markup in JavaScript, but compile JSX to a non-standard, highly optimized output.
+
+Instead of the `withHydrate` HOC, this compiler would be able to leverage a hydration directive directly inside of JSX. The compiler output would be highly optimized for SSR speed, relying on strings rather than a server-side VDOM.
+
+It's possible that hydration directives wouldn't even be necessary if we can detect whether a component uses `on*` event listeners or `use*` hooks. In that case, I'd default to `visible` hydration.
+
+```tsx
+// Input
+/* imports */
+
+const Input = () => {
+    return (
+        <>
+            <Head>
+                <seo.title>Microsite v2.0.0</seo.title>
+
+                <link rel="icon" type="image/svg+xml" href="/favicon.svg" />
+                <link rel="alternate icon" href="/favicon.ico" />
+            </Head>
+
+            <main>
+                <div class="stack">
+                    <h1>Hello world!</h1>
+                </div>
+
+                <Counter @hydrate={{ method: 'visible' }} />
+            </main>
+        </>
+    )
+}
+
+// Output
+import { ssr, component, escape } from 'microsite/internal';
+/* imports */
+
+const Output = () => {
+    return ssr`
+        ${component(Head, null, ssr`
+            ${component(seo.title, null, ssr`Microsite v2.0.0`)}
+            <link rel="icon" type="image/svg+xml" href="/favicon.svg" />
+            <link rel="alternate icon" href="/favicon.ico" />
+        `)}
+        <main>
+            <div class="stack">
+                <h1>Hello world!</h1>
+            </div>
+
+            ${component(Counter, { '@hydrate': { method: 'visible' } })}
+        </main>
+    `
+}
+```
+
+## Framework Agnostic
+
+Switching the majority of JSX compilation to a custom compiler would also open the door for other JSX-based frameworks. Preact would be the primary one, but Solid would be supported as well. 
+
+> I'm not interested in supporting React unless it is aliased to `preact/compat`.
+
+### Open Questions
+- Framework detection
+- Framework JSX compilation
+
+## Custom Rendering
+
+Microsite would have its own internal `renderToString` that resolves everything to a string of HTML.
+`component` would be able to render the component instance to a string of HTML and generate the final hydration script.
+
+### Streaming Rendering?
+
+We could potentially return a generator function to enable streaming responses by exposing something like `renderToStream`.
+
+## `esbuild`
+
+`esbuild` is pretty amazing, and there's also a Deno module. Assuming we could build an HMR engine for `esbuild`, do we really need all the extra weight that Snowpack/Vite bring along?
```

Comment:
👏👏👏

And while we’re on that topic, I think it might be good to be **very explicit** from the start about how v2 lives alongside the rest of the esbuild ecosystem. Which is to say, I think Microsite v2 should be designed to be a pipeable esbuild plugin, rather than a build step outside of or adjacent to esbuild.

Longer term, I’d like to also consider providing compiler APIs for additional AST transforms. This is currently a huge deoptimization/pain point with esbuild’s design. If you agree with this goal, I think it would be best to be ahead of the curve, by designing AST and transform as separate plugins even if they’re exposed as a single plugin by default.